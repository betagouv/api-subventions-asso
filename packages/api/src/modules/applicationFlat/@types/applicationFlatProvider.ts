import { FindCursor } from "mongodb";
import Provider from "../../providers/@types/IProvider";
import ApplicationFlatEntity from "../../../entities/ApplicationFlatEntity";

export default interface ApplicationFlatProvider extends Provider {
    isApplicationFlatProvider: true;

    findAllByExercise(exercise?: number): FindCursor;

    getApplicationFlatStream(START_YEAR: number, END_YEAR: number): ReadableStream<ApplicationFlatEntity>;
}
