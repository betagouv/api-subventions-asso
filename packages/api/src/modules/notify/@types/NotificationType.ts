export enum NotificationType {
    TEST_EMAIL,
    USER_CREATED,
    USER_DELETED,
    USER_UPDATED,
    USER_FORGET_PASSWORD,
}
