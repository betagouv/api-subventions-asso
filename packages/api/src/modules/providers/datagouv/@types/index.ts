export * from "./ISaveCallbacks";
export * from "./UniteLegalRaw";
