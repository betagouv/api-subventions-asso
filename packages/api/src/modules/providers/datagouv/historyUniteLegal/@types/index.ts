export * from "./ISaveCallbacks";
export * from "./UniteLegalHistoryRow"