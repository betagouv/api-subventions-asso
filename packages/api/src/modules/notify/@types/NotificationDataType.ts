import { NotificationType } from "./NotificationType";

export interface NotificationDataType {
    [NotificationType.USER_CREATED]: {
        email: string;
        token: string;
        signupAt: Date;
        active: boolean;
    };
    [NotificationType.USER_DELETED]: {
        email: string;
    };
    [NotificationType.USER_UPDATED]: {
        email: string;
        active: boolean;
    };
    [NotificationType.USER_FORGET_PASSWORD]: {
        email: string;
        token: string;
    };
    [NotificationType.TEST_EMAIL]: {
        email: string;
        templateId: number;
    };
}
