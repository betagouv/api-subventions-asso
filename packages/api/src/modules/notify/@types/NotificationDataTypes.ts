import { FutureUserDto, UserActivationInfoDto, UserDto } from "dto";
import { NotificationType } from "./NotificationType";

export interface NotificationDataTypes {
    [NotificationType.USER_ALREADY_EXIST]: {
        email: string;
        firstname?: string;
        lastname?: string;
        url?: string;
        signupAt: Date;
        active: boolean;
    };
    [NotificationType.USER_CREATED]: {
        email: string;
        firstname?: string;
        lastname?: string;
        url: string;
        signupAt: Date;
        active: boolean;
    };
    [NotificationType.USER_DELETED]: {
        email: string;
        firstname?: string;
        lastname?: string;
        selfDeleted: boolean;
    };
    [NotificationType.USER_ACTIVATED]: {
        email: string;
    };
    [NotificationType.USER_FORGET_PASSWORD]: {
        email: string;
        url: string;
    };
    [NotificationType.USER_LOGGED]: {
        email: string;
        date: Date;
    };
    [NotificationType.USER_UPDATED]: Omit<UserActivationInfoDto, "password"> & FutureUserDto;
    [NotificationType.USER_CONFLICT]: FutureUserDto;
    [NotificationType.SIGNUP_BAD_DOMAIN]: FutureUserDto;
    [NotificationType.TEST_EMAIL]: {
        email: string;
        templateId: number;
    };
    [NotificationType.FAILED_CRON]: {
        cronName: string;
        error: Error;
    };
}
