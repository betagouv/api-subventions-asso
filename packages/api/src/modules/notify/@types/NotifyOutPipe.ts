import { NotificationDataTypes } from "./NotificationDataTypes";
import { NotificationType } from "./NotificationType";

export interface NotifierMethodType {
    (type: NotificationType.TEST_EMAIL, data: NotificationDataTypes[NotificationType.TEST_EMAIL]): Promise<boolean>;
    (type: NotificationType.USER_DELETED, data: NotificationDataTypes[NotificationType.USER_DELETED]): Promise<boolean>;
    (type: NotificationType.USER_CREATED, data: NotificationDataTypes[NotificationType.USER_CREATED]): Promise<boolean>;
    (
        type: NotificationType.USER_ACTIVATED,
        data: NotificationDataTypes[NotificationType.USER_ACTIVATED],
    ): Promise<boolean>;
    (
        type: NotificationType.USER_FORGET_PASSWORD,
        data: NotificationDataTypes[NotificationType.USER_FORGET_PASSWORD],
    ): Promise<boolean>;
    (type: NotificationType.USER_LOGGED, data: NotificationDataTypes[NotificationType.USER_LOGGED]): Promise<boolean>;
    (
        type: NotificationType.USER_ALREADY_EXIST,
        data: NotificationDataTypes[NotificationType.USER_ALREADY_EXIST],
    ): Promise<boolean>;
    (type: NotificationType.USER_UPDATED, data: NotificationDataTypes[NotificationType.USER_UPDATED]);
    (type: NotificationType.SIGNUP_BAD_DOMAIN, data: NotificationDataTypes[NotificationType.SIGNUP_BAD_DOMAIN]);
    (type: NotificationType.FAILED_CRON, data: NotificationDataTypes[NotificationType.FAILED_CRON]);
}

export interface NotifyOutPipe {
    accepts: NotificationType[];
    notify: NotifierMethodType;
}
