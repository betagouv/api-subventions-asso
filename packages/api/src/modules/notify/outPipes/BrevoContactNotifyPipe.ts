import * as Sentry from "@sentry/node";
import Brevo from "@getbrevo/brevo";
import { AdminTerritorialLevel, AgentJobTypeEnum, AgentTypeEnum, TerritorialScopeEnum } from "dto";
import { NotificationDataTypes } from "../@types/NotificationDataTypes";
import { NotificationType } from "../@types/NotificationType";
import { NotifyOutPipe } from "../@types/NotifyOutPipe";
import { API_SENDINBLUE_CONTACT_LIST } from "../../../configurations/apis.conf";
import BrevoNotifyPipe from "./BrevoNotifyPipe";

const SENDIND_BLUE_CONTACT_LISTS = [Number(API_SENDINBLUE_CONTACT_LIST)];

/*
 * COMPTE_ACTIVE does not mean that the account is currently active, only that it has been through firstActivation
 * specifically, a user that had activated the account then lost they password, would have `active: false` on db
 * but COMPTE_ACTIVE: true on brevo */

export class BrevoContactNotifyPipe extends BrevoNotifyPipe implements NotifyOutPipe {
    private apiInstance: Brevo.ContactsApi;

    constructor() {
        super();
        this.apiInstance = new Brevo.ContactsApi();
    }

    notify(type, data) {
        switch (type) {
            case NotificationType.USER_CREATED:
                return this.userCreated(data);
            case NotificationType.USER_ALREADY_EXIST:
                return this.userCreated(data); // If user exist in datasub database so we create in brevo
            case NotificationType.USER_ACTIVATED:
                return this.userActivated(data);
            case NotificationType.USER_LOGGED:
                return this.userLogged(data);
            case NotificationType.USER_DELETED:
                return this.userDeleted(data);
            case NotificationType.BATCH_USERS_DELETED:
                return this.batchUsersDeleted(data);
            case NotificationType.USER_UPDATED:
                return this.userUpdated(data);
            default:
                return Promise.resolve(false);
        }
    }

    private userCreated(data: NotificationDataTypes[NotificationType.USER_CREATED]) {
        const payload = {
            attributes: {
                DATE_INSCRIPTION: data.signupAt,
                COMPTE_ACTIVE: data.active,
                SOURCE_IMPORT: "Data.Subvention",
                LIEN_ACTIVATION: data.url,
                PRENOM: data.firstname,
                NOM: data.lastname,
            },
            listIds: SENDIND_BLUE_CONTACT_LISTS,
        };

        const sendCreation = () => {
            return this.apiInstance.createContact({
                email: data.email,
                ...payload,
            });
        };

        const sendUpdate = () => this.apiInstance.updateContact(data.email, payload);

        return sendCreation()
            .then(({ body }) => {
                if (body?.id) return true;
                return false;
            })
            .catch(({ response, body: wrongBody }) => {
                // The true body is contained in response.body, body of error is wrong
                if (response.body.code === "duplicate_parameter") {
                    // It's error code for { code: 'duplicate_parameter', message: 'Contact already exist' }
                    // If user exists in other list brevo throws an error so we update contact to add in good list
                    return sendUpdate()
                        .then(updateAnwser => {
                            if (!updateAnwser.response) {
                                Sentry.captureException(updateAnwser);
                                return false;
                            }
                            const status = updateAnwser.response.statusCode || 0;
                            if (status >= 200 && status < 300) return true;
                            // Theoretically, this should never happen because mistakes should happen in "catch", but the lib seems to be behaving completely madly!
                            Sentry.captureException({ response: updateAnwser.response });
                            return false;
                        })
                        .catch(e => {
                            Sentry.captureException(e);
                            return false;
                        });
                }

                Sentry.captureException({ response, body: JSON.stringify(wrongBody) }); // Don't send to sentry if error is already caught
                return false;
            });
    }

    private userActivated(data: NotificationDataTypes[NotificationType.USER_ACTIVATED]) {
        const updateContact = new Brevo.UpdateContact();
        updateContact.attributes = {
            COMPTE_ACTIVE: true,
            LIEN_ACTIVATION: "",
        };
        updateContact.listIds = SENDIND_BLUE_CONTACT_LISTS;
        return this.apiInstance
            .updateContact(data.email, updateContact)
            .then(() => true)
            .catch(error => {
                Sentry.captureException(error);
                return false;
            });
    }

    private userLogged(data: NotificationDataTypes[NotificationType.USER_LOGGED]) {
        const updateContact = new Brevo.UpdateContact();
        updateContact.attributes = { DERNIERE_CONNEXION: data.date };
        updateContact.listIds = SENDIND_BLUE_CONTACT_LISTS;
        return this.apiInstance
            .updateContact(data.email, updateContact)
            .then(() => true)
            .catch(error => {
                Sentry.captureException(error);
                return false;
            });
    }

    private userDeleted(data: NotificationDataTypes[NotificationType.USER_DELETED]) {
        return this.apiInstance
            .deleteContact(data.email)
            .then(() => true)
            .catch(error => {
                Sentry.captureException(error);
                return false;
            });
    }

    private batchUsersDeleted(data: NotificationDataTypes[NotificationType.BATCH_USERS_DELETED]) {
        const promises = data.users.map(user =>
            this.apiInstance
                .deleteContact(user.email)
                .then(() => true)
                .catch(error => {
                    Sentry.captureException(error);
                    return false;
                }),
        );
        return Promise.all(promises).then(results => results.every(Boolean));
    }

    private userUpdated(data: NotificationDataTypes[NotificationType.USER_UPDATED]) {
        // TODO: get keys from a new shared DTO
        const JOB_TYPE_LABEL_MAPPING = {
            [AgentJobTypeEnum.ADMINISTRATOR]: "Gestionnaire administratif et financier",
            [AgentJobTypeEnum.EXPERT]: "Chargé de mission / Expert métier",
            [AgentJobTypeEnum.SERVICE_HEAD]: "Responsable de service",
            [AgentJobTypeEnum.CONTROLLER]: "Contrôleur / Inspecteur",
            [AgentJobTypeEnum.OTHER]: "Autre",
        };

        // TODO: get keys from a new shared DTO
        const AGENT_TYPE_LABEL_MAPPING = {
            [AgentTypeEnum.OPERATOR]: "Agent public d’un opérateur de l’État",
            [AgentTypeEnum.CENTRAL_ADMIN]: "Agent public d’une administration centrale (État)",
            [AgentTypeEnum.TERRITORIAL_COLLECTIVITY]: "Agent public d’une collectivité territoriale",
            [AgentTypeEnum.DECONCENTRATED_ADMIN]: "Agent public d’une administration déconcentrée (État)",
        };

        const ADMIN_TERRITORIAL_LEVEL_LABEL_MAPPING = {
            [AdminTerritorialLevel.DEPARTMENTAL]: "Départemental",
            [AdminTerritorialLevel.INTERDEPARTMENTAL]: "Interdépartemental",
            [AdminTerritorialLevel.REGIONAL]: "Régional",
            [AdminTerritorialLevel.INTERREGIONAL]: "Interrégional",
            [AdminTerritorialLevel.OVERSEAS]: "Collectivité d'outre-mer à statut particulier",
        };

        const TERRITORIAL_SCOPE_LABEL_MAPPING = {
            [TerritorialScopeEnum.REGIONAL]: "Régional",
            [TerritorialScopeEnum.DEPARTMENTAL]: "Départemental",
            [TerritorialScopeEnum.INTERCOMMUNAL]: "Intercommunal (EPCI)",
            [TerritorialScopeEnum.COMMUNAL]: "Communal",
            [TerritorialScopeEnum.OTHER]: "Autre",
        };

        const ATTRIBUTES_MAPPING = {
            agentType: "CATEGORIE",
            service: "SERVICE",
            phoneNumber: "TELEPHONE",
            jobType: "BUYER_PERSONNAE",
            structure: "STRUCTURE",
            decentralizedLevel: "COMPETENCE_TERRITORIALE",
            decentralizedTerritory: "ECHELON_TERRITORIAL",
            territorialScope: "ECHELON_COLLECTIVITE",
            lastName: "NOM",
            firstName: "PRENOM",
            lastActivityDate: "DERNIERE_CONNEXION",
        };

        function buildAttributesObject(data) {
            return Object.keys(data).reduce((acc, key) => {
                switch (key) {
                    case "agentType":
                        acc[ATTRIBUTES_MAPPING[key]] = AGENT_TYPE_LABEL_MAPPING[data[key]];
                        break;
                    case "jobType":
                        acc[ATTRIBUTES_MAPPING[key]] = (data[key] || [])
                            .map(job => JOB_TYPE_LABEL_MAPPING[job])
                            .join(",");
                        break;
                    case "decentralizedLevel":
                        acc[ATTRIBUTES_MAPPING[key]] = ADMIN_TERRITORIAL_LEVEL_LABEL_MAPPING[data[key]];
                        break;
                    case "territorialScope":
                        acc[ATTRIBUTES_MAPPING[key]] = TERRITORIAL_SCOPE_LABEL_MAPPING[data[key]];
                        break;
                    default:
                        if (Object.keys(ATTRIBUTES_MAPPING).includes(key)) acc[ATTRIBUTES_MAPPING[key]] = data[key];
                }

                if (typeof acc[ATTRIBUTES_MAPPING[key]] == "string" && !acc[ATTRIBUTES_MAPPING[key]].length) {
                    delete acc[ATTRIBUTES_MAPPING[key]];
                }
                return acc;
            }, {});
        }

        const updateContact = new Brevo.UpdateContact();
        const attributes: Record<string, string | boolean> = buildAttributesObject(data);
        attributes.COMPTE_ACTIVE = true;
        updateContact.attributes = attributes;
        updateContact.listIds = SENDIND_BLUE_CONTACT_LISTS;
        return this.apiInstance
            .updateContact(data.email, updateContact)
            .then(() => true)
            .catch(error => {
                Sentry.captureException(error);
                console.error("error updating contact", { email: data.email, error: error.response._body });
                return false;
            });
    }
}

const brevoContactNotifyPipe = new BrevoContactNotifyPipe();

export default brevoContactNotifyPipe;
