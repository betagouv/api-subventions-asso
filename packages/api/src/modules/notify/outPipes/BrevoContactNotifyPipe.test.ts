import { API_SENDINBLUE_CONTACT_LIST } from "../../../configurations/apis.conf";
import { NotificationType } from "../@types/NotificationType";
import BrevoContactNotifyPipe from "./BrevoContactNotifyPipe";

describe("BrevoContactNotifyPipe", () => {
    const USER_EMAIL = "user@beta.gouv.fr";

    // @ts-expect-error apiInstance is private attribute
    const mockUpdateContact = jest.spyOn(BrevoContactNotifyPipe.apiInstance, "updateContact");

    beforeAll(() => {
        mockUpdateContact.mockResolvedValue({ body: { id: 1 } } as any);
    });

    describe("notify", () => {
        let userCreatedSpy: jest.SpyInstance;

        beforeAll(() => {
            // @ts-expect-error userCreated is private methode
            userCreatedSpy = jest.spyOn(BrevoContactNotifyPipe, "userCreated").mockResolvedValue(true);
        });

        afterAll(() => {
            userCreatedSpy.mockRestore();
        });

        it("should call userCreated", () => {
            BrevoContactNotifyPipe.notify(NotificationType.USER_CREATED, {});

            expect(userCreatedSpy).toBeCalled();
        });

        it("should not call userCreated", () => {
            BrevoContactNotifyPipe.notify(NotificationType.TEST_EMAIL, {});

            expect(userCreatedSpy).not.toBeCalled();
        });
    });

    describe("userCreated", () => {
        let createContactSpy: jest.SpyInstance;

        beforeAll(() => {
            createContactSpy = jest
                // @ts-expect-error apiInstance is private attribute
                .spyOn(BrevoContactNotifyPipe.apiInstance, "createContact")
                .mockResolvedValue({ body: { id: 1 } } as any);
        });

        afterAll(() => {
            createContactSpy.mockRestore();
        });

        it("should call create data", async () => {
            const expected = {
                email: USER_EMAIL,
                attributes: {
                    DATE_INSCRIPTION: new Date(),
                    COMPTE_ACTIVE: true,
                    SOURCE_IMPORT: "Data.Subvention",
                    LIEN_ACTIVATION: "TOKEN",
                },
                listIds: [Number(API_SENDINBLUE_CONTACT_LIST)],
            };

            // @ts-expect-error userCreated is private method
            await BrevoContactNotifyPipe.userCreated({
                email: expected.email,
                signupAt: expected.attributes.DATE_INSCRIPTION,
                active: expected.attributes.COMPTE_ACTIVE,
                token: expected.attributes.LIEN_ACTIVATION,
            });

            expect(createContactSpy).toHaveBeenCalledWith(expected);
        });
    });

    describe("userActivated", () => {
        afterEach(() => {
            mockUpdateContact.mockClear();
        });

        it("call updateContact()", async () => {
            // @ts-expect-error userCreated is private method
            await BrevoContactNotifyPipe.userActivated({ email: USER_EMAIL });
            expect(mockUpdateContact).toHaveBeenCalledWith(USER_EMAIL, {
                attributes: { COMPTE_ACTIVE: true },
                listIds: [Number(API_SENDINBLUE_CONTACT_LIST)],
            });
        });
    });

    describe("userLogged", () => {
        afterEach(() => {
            mockUpdateContact.mockClear();
        });

        it("should call updateContact()", async () => {
            const email = "UserEmail";
            const now = new Date();

            // @ts-expect-error userLogged is private method
            await BrevoContactNotifyPipe.userLogged({ email, date: now });

            expect(mockUpdateContact).toHaveBeenCalledWith(email, {
                attributes: { DERNIERE_CONNEXION: now },
                listIds: [Number(API_SENDINBLUE_CONTACT_LIST)],
            });
        });
    });
});
