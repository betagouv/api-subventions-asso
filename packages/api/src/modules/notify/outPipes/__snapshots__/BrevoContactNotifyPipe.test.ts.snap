// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrevoContactNotifyPipe batchUsersDeleted should call deleteContact() for each user 1`] = `
Array [
  Array [
    "test@datasubvention.gouv.fr",
  ],
  Array [
    "test2@datasubvention.gouv.fr",
  ],
]
`;

exports[`BrevoContactNotifyPipe userDeleted should call deleteContact() 1`] = `
Array [
  "test@datasubvention.gouv.fr",
]
`;

exports[`BrevoContactNotifyPipe userUpdated should call updateContact() 1`] = `
Array [
  "test@datasubvention.gouv.fr",
  exports {
    "attributes": Object {
      "BUYER_PERSONNAE": "Gestionnaire administratif et financier,Chargé de mission / Expert métier",
      "CATEGORIE": "Agent public d’une collectivité territoriale",
      "COMPETENCE_TERRITORIALE": "Interdépartemental",
      "COMPTE_ACTIVE": true,
      "ECHELON_COLLECTIVITE": "Départemental",
      "ECHELON_TERRITORIAL": "ECHELON_TERRITORIAL",
      "NOM": "NOM",
      "PRENOM": "PRENOM",
      "SERVICE": "SERVICE",
      "TELEPHONE": "+33625859685",
    },
    "listIds": Array [
      1,
    ],
  },
]
`;
